File: mobile/lib/core/utils/email_validator.dart
class EmailValidator {
  static final RegExp _localPartPattern =
      RegExp(r"^[A-Za-z0-9.!#$%&'*+/=?^_`{|}~-]+$");
  static final RegExp _domainLabelPattern = RegExp(r'^[A-Za-z0-9-]+$');
  static final RegExp _tldPattern = RegExp(r'^[A-Za-z]{2,63}$');

  static bool isValid(String value) {
    final email = value.trim().toLowerCase();

    if (email.isEmpty || email.length > 254 || email.contains(' ')) {
      return false;
    }

    final List<String> parts = email.split('@');
    if (parts.length != 2) {
      return false;
    }

    final String localPart = parts[0];
    final String domain = parts[1];

    if (localPart.isEmpty ||
        domain.isEmpty ||
        localPart.length > 64 ||
        localPart.startsWith('.') ||
        localPart.endsWith('.') ||
        localPart.contains('..') ||
        !_localPartPattern.hasMatch(localPart)) {
      return false;
    }

    if (domain.startsWith('.') ||
        domain.endsWith('.') ||
        domain.contains('..')) {
      return false;
    }

    final List<String> labels = domain.split('.');
    if (labels.length < 2) {
      return false;
    }

    for (final String label in labels) {
      if (label.isEmpty ||
          label.length > 63 ||
          label.startsWith('-') ||
          label.endsWith('-') ||
          !_domainLabelPattern.hasMatch(label)) {
        return false;
      }
    }

    return _tldPattern.hasMatch(labels.last);
  }
}

File: services/email-validator.js
const LOCAL_PART_PATTERN = /^[A-Za-z0-9.!#$%&'*+/=?^_`{|}~-]+$/;
const DOMAIN_LABEL_PATTERN = /^[A-Za-z0-9-]+$/;
const TLD_PATTERN = /^[A-Za-z]{2,63}$/;

function isValidEmail(value) {
  const email = String(value || "").trim().toLowerCase();

  if (!email || email.length > 254 || email.includes(" ")) {
    return false;
  }

  const parts = email.split("@");
  if (parts.length !== 2) {
    return false;
  }

  const [localPart, domain] = parts;

  if (
    !localPart ||
    !domain ||
    localPart.length > 64 ||
    localPart.startsWith(".") ||
    localPart.endsWith(".") ||
    localPart.includes("..") ||
    !LOCAL_PART_PATTERN.test(localPart)
  ) {
    return false;
  }

  if (
    domain.startsWith(".") ||
    domain.endsWith(".") ||
    domain.includes("..")
  ) {
    return false;
  }

  const labels = domain.split(".");
  if (labels.length < 2) {
    return false;
  }

  for (const label of labels) {
    if (
      !label ||
      label.length > 63 ||
      label.startsWith("-") ||
      label.endsWith("-") ||
      !DOMAIN_LABEL_PATTERN.test(label)
    ) {
      return false;
    }
  }

  return TLD_PATTERN.test(labels[labels.length - 1]);
}

module.exports = {
  isValidEmail,
};
